### DISCUSSION TITLE
Conditionally run jobs based on file changes

### DISCUSSION THREAD

**Discussion by abstractj** [keycloak/keycloak#13066] [Ideas]
## Description

Every time we create a new pull-request, GitHub actions will trigger approximately 28 checks, taking approximately 6 hours to complete.

There are some scenarios where the changes proposed only touches a couple of JavaScript, NPM dependencies or documentation files. We should be able to run a subset of jobs in the pipeline based on those changes.

Something similar to the current approach for the workflows [CodeQL JS Adapter](https://github.com/keycloak/keycloak/blob/main/.github/workflows/codeql-js-adapter-analysis.yml) and [CodeQL Themes](https://github.com/keycloak/keycloak/blob/main/.github/workflows/codeql-theme-analysis.yml).

## Proposal

1. Themes

For JavaScript related changes and NPM dependencies, we can conditionally execute the following jobs only based on the changes inside the `themes` folder:  
- CodeQL Themes / CodeQL analyze  
- Keycloak CI / Build (pull_request)  
- Keycloak CI / Quickstarts Tests

A concrete example is available in the [Update TypeScript to the latest version](https://github.com/keycloak/keycloak/pull/13060). Even though the changes proposed only touches a typescript file and `package.json`, jobs for the Operator and Cross DC are executed anyway. That means 6 hours only for a couple of changes.

2. Docs

For changes to the docs, I don't see any need to execute workflows for the Operator, Cross DC, WebAuthN…or unit tests. If there's any need to build and lint our doc, we should be able to run only the relevant workflow.

The changes proposed in the pull-request [Fix typo](https://github.com/keycloak/keycloak/pull/12896/files) triggers 27 checks, only for a couple of docs changes.

If necessary, we can run all the workflows on the daily basis to make sure that nothing was broken.

## Future

In the future, we may consider expanding the scope, split the workflow files and conditionally trigger the jobs based on folders:

-   Operator
-   Quarkus
-   SAML*
-   Authz
-   ...

## Related posts

-   [Improve CI setup](https://github.com/keycloak/keycloak/discussions/9947)

## Warning

This is not the perfect solution and open for suggestions. The goal here is to iterate over it, considering that speed is something crucial for our developer experience.

**Comment by abstractj** [comment]
cc @keycloak/kc-developers

**Comment by DGuhr** [comment]
I went down the path some time in the past when we added the new guides, that also led to severe CI resource exhaustion. IIRC there were some bumps hindering us to go further with the "conditionally execute stuff based on files touched" approach:
* we have branch protection on main, and some actions are marked as `required` before a merge is possible.
* This effectively could leave an open PR in an "unmergeable" state, because the required CI steps would never run when we leave them out on specific conditions.
* GitHub Actions does not provide proper support for this scenario, yet.
* Ref to old GitHub discussions:
  * https://github.community/t/feature-request-conditional-required-checks/16761
  * https://github.com/github-community/community/discussions/13690 (seems to be the follow-up of the link above, have to scan through that one myself, just found it this morning)

Example: for a docs-only change, e.g. a typo, a (simplified) idea was to have something around the lines of:

```
on:
  push:
    paths-ignore:
      - 'docs/guides/**'
  pull_request:
    paths-ignore:
      - 'docs/guides/**'
```
and a specific job that just runs only the steps needed to see if the guides can be parsed correctly. Now our matrix testsuite is required, but not executed. This leads to an unmergeable state for a purely docs-related PR. 

That being said, consider this just a point of information why I stopped looking into improving it back then. The idea makes total sense, I have not keepen track of the latest developments around this topic on the GHa side and hope it is somehow possible to achieve this.

**edit**: the [github discussions follow-up](https://github.com/github-community/community/discussions/13690) links to a documentation bit that states that it is possible with a workaround (for now) that is having a "return true" dummy workflow with the same name and the inverse filtering logic. This feels very hacky and would pollute our CI script quite a bit, but might work unless there's better support. In the meantime, feel free to give the thread a vote, so hopefully we will have better support in the future.

**Reply by DGuhr** [reply]
I remember, @andreaTP had some ideas around this after talking to the quarkus team, too. Would be nice to hear them :)

**Comment by ahus1** [comment]
Some more options (maybe not solutions, yet): 

* Some parts like the "test-cluster" already look into the most recent git diff to figure out if they should run. If not, they complete early, see GIT_DIFF here: 
https://github.com/keycloak/keycloak/blob/a1c0e5b11334a5038841de86e3fe38076214f7a2/.github/workflows/ci.yml#L295-L313
This could be a workaround for the problem with the required actions that @DGuhr describes.

* There is the Maven extension / plugin https://github.com/gitflow-incremental-builder/gitflow-incremental-builder that is used for example in Quarkus. Despite its name "gitflow", it allows incremental builds for Maven projects by looking at the changes committed in a given branch. It might be able to determine the dependencies between modules where we don't want to run some of the tests.

* Using the gitflow-incremental-builder it might be possible to speed up the Unit Test build as follows (it's a one-liner once that maven extension is active):
   * identify all changes modules
   * build all upstream modules for the changed modules, but skip the tests for those
   * build all changed modules and the downstream modules and run the unit tests on those
 
  Depending on the results, one could even skip downloading the pre-built keycloak which takes ~2 min to download.

BTW: I'd assume the gitflow-incremental-builder comes in quite handy when working on the Quarkus distribution of Keycloak, as one would regularly build the non-Quarkus-modules. Maybe someone with Quarkus development experience can comment on that.

**Reply by andreaTP** [reply]
@ahus1 you anticipated me!

Those are the technical information I gathered from the super kind @gsmet ( thanks again! 🙏 ) when looking at how to integrate GIB:

```
Incremental build

The initial PR with the incremental build that only builds Maven modules that are actually affected by a change is here: https://github.com/quarkusio/quarkus/pull/13243 .

It has evolved since, here are a few pointers:
- Setup in the parent pom: https://github.com/quarkusio/quarkus/blob/main/pom.xml#L315
- How to get the impacted modules: https://github.com/quarkusio/quarkus/blob/main/.github/workflows/ci-actions-incremental.yml#L180 and know if a job is needed
- But it's also run when actually running the tests here: https://github.com/quarkusio/quarkus/blob/main/.github/workflows/ci-actions-incremental.yml#L341-L342

Load matrix from JSON

Here is the trick to load the matrix from a JSON file: https://github.com/quarkusio/quarkus/blob/main/.github/workflows/ci-actions-incremental.yml#L667 .

In this case, we load the JSON from a previous step output but you can load it from a JSON file that is in your repo too by giving a path like we did in our old workflow:
https://github.com/quarkusio/quarkus/blob/main/.github/workflows/ci-actions.yml.disabled#L517
```

**Comment by abstractj** [comment]
Thank you, @DGuhr @ahus1 and @andreaTP for the feedback. Those are interesting ideas to explore. As short-term solution and an immediate step. Would make sense to have a PR based on the simplified idea suggested by @DGuhr ? I don't think it will solve all the issues, but it's a small and important improvement from my perspective, and we can iterate over it.

For now, I'm thinking about prevent our workflows to run on pull-requests for docs and translation files. Something like:

```
on:
  pull_request:
    paths-ignore:
      - 'docs/**' (we don't need to run tests when a markdown file changes)
      - 'themes/**/*.properties' (from what I checked we don't have integration tests for our translations. That means no tests for pt_BR messages for example.)
      - 'examples/**' (those are old examples, eventually updated but can run as part of the nightly jobs)
      - 'adapters/**/jetty/**' (OIDC and SAML adapters were deprecated and can run as part of the nightly jobs)
      - 'adapters/**/tomcat/**' (OIDC and SAML adapters were deprecated and can run as part of the nightly jobs)
      - 'adapters/adapters/oidc/wildfly/**' (deprecated and can run as part of the nightly jobs)
```

As a second step, explore the options suggested with `GIT_DIFF`, and also the approach from Quarkus. Does it make any sense?

**Reply by andreaTP** [reply]
Preventing CI on docs and translations is a good low hanging fruit.
Although I'm not fond into the `GIT_DIFF` approach as it's not a good enough foundation and will require continuous tweaks and iterations as the project evolve while the Quarkus approach is more principled, robust, safe and proven to be effective.

**Reply by mposolda** [reply]
I would be careful with adding `themes/**/*.properties` as some tests rely on the text displayed on the page or expected error message displayed on the page etc and it can easily happen that some test will be broken due the change in the message. And AFAIK there are not so much PRs, which affect only property files?

I also hope that long CI run will be "automagically fixed" by removing Wildfly distribution, which will allow to remove all "Base Testsuite Wildfly" build, which are long and time consuming to run. But agree that at the same time, we can try to explore better ways of conditionally removing jobs instead of relying on "git diff" like we do actually.

**Reply by DGuhr** [reply]
In addition to what marek states, it is the same with 'docs/**' as there's also the maven plugin inside and a freemarker layer. There are no tests, but at least compilation should always be tested. better: `docs/**/*.adoc`

**Reply by DGuhr** [reply]
For the second step, I would also prefer to look into the battle-proven Quarkus approach before going down the git diff road, for the reasons @andreaTP mentioned, and because there's already a hacky workaround (see link in my previous post)

**Reply by vmuzikar** [reply]
I think conditional runs will always be a bit tricky due to how tightly coupled the repo is. And that's natural – the tightly coupled components are in a single repo for a reason.

Yet I still believe conditional test runs are feasible. We just need to make sure we run the full test workflow on daily basis and any missed regressions are fixed ASAP. Ideally, we should resurrect the Merge Monster idea for that...
